objects/src/lib.rs

```diff
@@ -44,79 +44,94 @@ pub mod crypto {
+        let mut felts = [0u64; 4];
+        let mut i = 0;
+        while i < hex_bytes.len() {
+            let hex_digit = match parse_hex_digit(hex_bytes[i]) {
+                Ok(v) => v as u64,
+                Err(e) => return Err(e),
             };
 
-            if hex_bytes.len() > 64 {
-                panic!("Hex string has more than 64 characters");
-            }
+            // This digit's nibble offset within the felt. We need to invert the nibbles per
+            // byte for endianess reasons i.e. ABCD -> BADC.
+            let inibble = if i % 2 == 0 { (i + 1) % 16 } else { (i - 1) % 16 };
+            let value = hex_digit << (inibble * 4);
+            felts[i / 2 / 8] += value;
 
-            // Aggregate each byte into the appropriate felt value. Doing it this way also allows
-            // for variable string lengths.
-            let mut felts = [0u64; 4];
-            let mut i = 0;
-            // We are forced to use a while loop because the others aren't supported in const
-            // context.
-            while i < hex_bytes.len() {
-                // This digit's nibble offset within the felt. We need to invert the nibbles per
-                // byte for endianess reasons i.e. ABCD -> BADC.
-                let inibble = if i % 2 == 0 { (i + 1) % 16 } else { (i - 1) % 16 };
-
-                // SAFETY: u8 cast to u64 is safe. We cannot use u64::from in const context so we
-                // are forced to cast.
-                let value = (parse_hex_digit(hex_bytes[i]) as u64) << (inibble * 4);
-                felts[i / 2 / 8] += value;
-
-                i += 1;
-            }
+            i += 1;
+        }
 
-            // Ensure each felt is within bounds as `Felt::new` silently wraps around.
-            // This matches the behaviour of `Digest::try_from(String)`.
-            let mut i = 0;
-            while i < felts.len() {
-                use $crate::StarkField;
-                if felts[i] > $crate::Felt::MODULUS {
-                    panic!("Felt overflow");
-                }
-                i += 1;
+        // Ensure each felt is within bounds as `Felt::new` silently wraps around.
+        // This matches the behaviour of `Digest::try_from(String)`.
+        let mut idx = 0;
+        while idx < felts.len() {
+            if felts[idx] > Felt::MODULUS {
+                return Err("Felt overflow");
             }
+            idx += 1;
+        }
+
+        Ok([
+            Felt::new(felts[0]),
+            Felt::new(felts[1]),
+            Felt::new(felts[2]),
+            Felt::new(felts[3]),
+        ])
+    }
+
+    /// Construct a new `Digest` from a hex value.
+    ///
+    /// Expects a '0x' prefixed hex string followed by up to 64 hex digits.
+    #[macro_export]
+    macro_rules! digest {
+        ($hex:expr) => {{
+            let felts: [$crate::Felt; 4] = match $crate::utils::parse_hex_string_as_word($hex) {
```

Comment:
Question: how would this affect the ability to use the `digest!` macro from other crates?